src/data_access/JSONPersistence.java

```diff
@@ -0,0 +1,155 @@
+            array.put(recipe.getMealId());
+        }
+        return array;
+    }
+
+    /**
+     * Create a JSONObject that contains the user's tags (with associated
+     * recipes) and favourites.
+     * @param user The User object on which we base the output
+     * @return A JSONObject that has two keys: "favourites" and "tags".
+     * The value for "favourites" is a JSONArray of the IDs of the
+     * "favourited" recipes. The value for "tags" is a JSONObject, which
+     * contains the tag names as keys, and JSONArrays of the IDs of the
+     * recipes as the values.
+     */
+    private JSONObject constructJSONObjectFromUser(User user) {
+        JSONObject obj = new JSONObject();
+        JSONObject tagsJSON = JSONForUserTags(user);
+        JSONArray favouritesJSON = JSONForUserFavourites(user);
+        obj.put("username", user.getUsername());
+        obj.put("tags", tagsJSON);
+        obj.put("favourites", favouritesJSON);
+        return obj;
+    }
+
+    /**
+     * Save the user's tags and favourites to the file that was given upon
+     * object creation.
+     * @param user the User object whose contents are saved.
+     */
+    public void save(User user) {
+        JSONObject obj = constructJSONObjectFromUser(user);
+        BufferedWriter writer;
+        try {
+            writer = new BufferedWriter(new FileWriter(this.filePath));
+            writer.write(obj.toString(4));
+            writer.close();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public User load() {
+        User user;
+        String name;
+        JSONArray favourites;
+        JSONObject tagsMap;
+        /* Read and parse the information from the file. */
+        try {
+            String fileContents = Files.readString(Path.of(this.filePath));
+            JSONTokener tokenizer = new JSONTokener(fileContents);
+            JSONObject obj = new JSONObject(tokenizer);
+            name = obj.getString("username");
+            favourites = obj.getJSONArray("favourites");
+            tagsMap = obj.getJSONObject("tags");
+        } catch (JSONException | IOException e) {
+            throw new RuntimeException(e);
+        }
+        /* Construct the user object. */
+        /* TODO: fix this hack, where we don't use the password. */
+        user = userFactory.create(name, "");
+        /* Update user with favourites */
+        Recipe[] favouritesList = generateRecipesFromJSONArray(favourites);
+        for (Recipe recipe : favouritesList) {
+            user.addFavourite(recipe);
+        }
+        /* Update user with tags */
+        for (String key: tagsMap.keySet()) {
+            Recipe[] recipeArray = generateRecipesFromJSONArray(tagsMap.getJSONArray(key));
+            for (Recipe recipe : recipeArray) {
+                user.assignTag(recipe, key);
+            }
+        }
+        return user;
+    }
+
+    private Recipe[] generateRecipesFromJSONArray(JSONArray array) {
```

Comment:
A more efficient way to query a list of IDs and get back a list of Recipe objects is a great next step, perhaps something we could work on together because this seems interesting to me.